operators/distiller-pipeline/distiller_pipeline/__init__.py

```diff
@@ -0,0 +1,981 @@
+from enum import Enum
+from typing import Any
+
+import msgpack
+import pandas as pd
+import zmq
+from pydantic import BaseModel, Field, ValidationError
+
+
+# --- Enum Mappings ---
+class StateLocation:
+    RECEIVER = 0
+    AGGREGATOR = 1
+    NODE = 2
+
+
+class MessagingStatus:
+    IDLE = 0
+    BEFORE_SEND_FRAMES = 1
+    AFTER_SEND_FRAMES = 2
+    WAITING_FOR_UPSTREAM_FRAMES = 3
+    WAITING_FOR_DOWNSTREAM_READY_FRAMES = 4
+    WAITING_FOR_DOWNSTREAM_READY_MULTISCAN = 5
+    STREAMING = 6
+    FINISHED_SEND = 7
+    PROCESSING = 8
+    WRITING_TO_PM = 9
+    FINISHED_WRITING_TO_PM = 10
+    WRITING = 11
+    PULLING_FROM_FPGA = 12
+
+
+def state_location_to_string(location_int: int) -> str:
+    mapping = {
+        StateLocation.RECEIVER: "RECEIVER",
+        StateLocation.AGGREGATOR: "AGGREGATOR",
+        StateLocation.NODE: "NODE",
+    }
+    return mapping.get(location_int, "UNKNOWN")
+
+
+def messaging_status_to_string(status_int: int) -> str:
+    mapping = {
+        MessagingStatus.IDLE: "Idle",
+        MessagingStatus.BEFORE_SEND_FRAMES: "B-Frames",
+        MessagingStatus.AFTER_SEND_FRAMES: "A-Frames",
+        MessagingStatus.WAITING_FOR_UPSTREAM_FRAMES: "WU-Frames",
+        MessagingStatus.WAITING_FOR_DOWNSTREAM_READY_FRAMES: "WD-FRAMES",
+        MessagingStatus.WAITING_FOR_DOWNSTREAM_READY_MULTISCAN: "WD-Multi",
+        MessagingStatus.STREAMING: "Streaming",
+        MessagingStatus.FINISHED_SEND: "Fin-Send",
+        MessagingStatus.PROCESSING: "Processing",
+        MessagingStatus.WRITING_TO_PM: "WritingPM",
+        MessagingStatus.FINISHED_WRITING_TO_PM: "FinWritePM",
+        MessagingStatus.WRITING: "Writing",
+        MessagingStatus.PULLING_FROM_FPGA: "Pulling",
+    }
+    return mapping.get(status_int, "UNKNOWN")
+
+
+# --- Pydantic Models (Mirroring C++ Structs) ---
+
+
+class ReceiverInfo(BaseModel):
+    status: int = MessagingStatus.IDLE
+    module: int = 0
+    n_messages: int = 0
+    scan_number: int = 0
+
+    @classmethod
+    def from_msgpack(cls, data: list[Any]) -> "ReceiverInfo":
+        if isinstance(data, list) and len(data) >= 4:
+            try:
+                return cls(
+                    status=data[0],
+                    module=data[1],
+                    n_messages=data[2],
+                    scan_number=data[3],
+                )
+            except (ValidationError, ValueError, TypeError):
+                pass
+
+        return cls()
+
+
+class ReceiversInfo(BaseModel):
+    n_messages: int = 0
+    receivers: dict[int, dict[int, ReceiverInfo]] = Field(default_factory=dict)
+
+    @classmethod
+    def from_msgpack(cls, data: list[Any]) -> "ReceiversInfo":
+        instance_data: dict[str, Any] = {
+            "n_messages": 0,
+            "receivers": {},
+        }
+        if isinstance(data, list) and len(data) >= 2:
+            try:
+                instance_data["n_messages"] = int(data[0])
+            except (ValueError, TypeError):
+                pass
```

Comment:
These value/type errors can be ignored? In what situation could these be raised?